### ISSUE TITLE
taint resources in a module with count

### ISSUE THREAD

**Issue by issacg** [hashicorp/terraform#25535]
### Terraform Version

```
0.13.0-beta2
```

### Expected Behavior

I have a module with a submodule that has count on it, as part of testing 0.13 to see how it can make my life more awesome

I can create the resources quite easily 

```
terraform apply
...
module.cm-test-instances.module.web[1].aws_instance.web: Creating...
module.cm-test-instances.module.web[2].aws_instance.web: Creating...
module.cm-test-instances.module.web[0].aws_instance.web: Creating...
...
```

I expect to be able to taint the resources via, say, `terraform taint module.cm-test-instances.module.web[1].aws_instance.web`

(bonus points if there is syntax so that I could taint all of them at once via the cli)
### Actual Behavior

```
Error: No such resource instance

The state currently contains no resource instances whatsoever. This may occur
if the configuration has never been applied or if it has recently been
destroyed.
```

**Comment by jbardin** [comment]
Hi @issacg,

Square brackets are usually a reserved character in a shell, and often a binary as well; e.g.:

```
% type -a [
[ is a shell builtin
[ is '/bin/['
```

So what you see on the command line is not likely what is passed into terraform. For most unix shells, you can quote the argument to interpret literally:

```
terraform taint 'module.cm-test-instances.module.web[1].aws_instance.web'
```

As for matching multiple instances with the taint command, we have an open issue here (#1768) we are using to track that feature.

Thanks!

**Comment by issacg** [comment]
I'm pretty sure it's not shell escaping, because when I do `terraform taint module.cm-test-instances.module.web[1].aws_instance`

I get 
```
Error: Invalid address

  on  line 1:
  (source code not available)

Resource specification must include a resource type and name.
```

In any case, neither quoting the string nor attempting to escape the brackets with `\`'s seems to make a difference :(

`terraform taint 'module.cm-test-instances.module.web[1].aws_instance.web'` and `terraform taint module.cm-test-instances.module.web\[1\].aws_instance.web` both output

```
Error: No such resource instance

The state currently contains no resource instances whatsoever. This may occur
if the configuration has never been applied or if it has recently been
destroyed.
```

**Comment by jbardin** [comment]
The second error appears to be that there are no resources in the state at all, but that would be unrelated to parsing a resource address.

The first error appears that terraform is getting a malformed address to parse. Running the command with `TF_LOG=trace` should show the exact parameters that terraform is getting from the shell.

**Comment by issacg** [comment]
Sorry for the delayed response, @jbardin.  I'm still scratching my head over this :/

The state definitely exists:
```
vscode@ab5e4392765f:/workspaces/terraform/cm-sandbox$ terraform state show module.cm-test-instances.module.web[1].aws_instance.web
# module.cm-test-instances.module.web[1].aws_instance.web:
resource "aws_instance" "web" {
    ami                          = "ami-0b1e2eeb33ce3d66f"
    [...snip...]
}
```

It's intuitively obvious that this isn't a shell issue, or that wouldn't work.  Setting the log to `trace` didn't shed any immediate light on anything (beyond that it's not shell parsing)
```
vscode@ab5e4392765f:/workspaces/terraform/cm-sandbox$ TF_LOG=trace terraform taint module.cm-test-instances.module.web[1].aws_instance.web
2020/07/15 11:25:34 [INFO] Terraform version: 0.13.0 beta3 
2020/07/15 11:25:34 [INFO] Go runtime version: go1.14.2
2020/07/15 11:25:34 [INFO] CLI args: []string{"/usr/local/bin/terraform", "taint", "module.cm-test-instances.module.web[1].aws_instance.web"}
2020/07/15 11:25:34 [DEBUG] Attempting to open CLI config file: /home/vscode/.terraformrc
2020/07/15 11:25:34 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/07/15 11:25:34 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2020/07/15 11:25:34 [DEBUG] ignoring non-existing provider search directory /home/vscode/.terraform.d/plugins
2020/07/15 11:25:34 [DEBUG] ignoring non-existing provider search directory /home/vscode/.local/share/terraform/plugins
2020/07/15 11:25:34 [DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins
2020/07/15 11:25:34 [DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins
2020/07/15 11:25:34 [INFO] CLI command args: []string{"taint", "module.cm-test-instances.module.web[1].aws_instance.web"}
2020/07/15 11:25:34 [TRACE] Meta.Backend: no config given or present on disk, so returning nil config
2020/07/15 11:25:34 [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2020/07/15 11:25:34 [DEBUG] New state was assigned lineage "f3ff8551-67fc-696b-d570-b3e951738451"
2020/07/15 11:25:34 [TRACE] Meta.Backend: using default local state only (no backend configuration, and no existing initialized backend)
2020/07/15 11:25:34 [TRACE] Meta.Backend: instantiated backend of type <nil>
2020/07/15 11:25:34 [TRACE] providercache.fillMetaCache: scanning directory .terraform/plugins
2020/07/15 11:25:34 [TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/aws v2.69.0 for linux_amd64 at .terraform/plugins/registry.terraform.io/hashicorp/aws/2.69.0/linux_amd64
2020/07/15 11:25:34 [TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/aws v2.70.0 for linux_amd64 at .terraform/plugins/registry.terraform.io/hashicorp/aws/2.70.0/linux_amd64
2020/07/15 11:25:34 [TRACE] getproviders.SearchLocalDirectory: found registry.terraform.io/hashicorp/tls v2.1.1 for linux_amd64 at .terraform/plugins/registry.terraform.io/hashicorp/tls/2.1.1/linux_amd64
2020/07/15 11:25:34 [TRACE] providercache.fillMetaCache: including .terraform/plugins/registry.terraform.io/hashicorp/aws/2.69.0/linux_amd64 as a candidate package for registry.terraform.io/hashicorp/aws 2.69.0
2020/07/15 11:25:34 [TRACE] providercache.fillMetaCache: including .terraform/plugins/registry.terraform.io/hashicorp/aws/2.70.0/linux_amd64 as a candidate package for registry.terraform.io/hashicorp/aws 2.70.0
2020/07/15 11:25:34 [TRACE] providercache.fillMetaCache: including .terraform/plugins/registry.terraform.io/hashicorp/tls/2.1.1/linux_amd64 as a candidate package for registry.terraform.io/hashicorp/tls 2.1.1
2020/07/15 11:25:35 [TRACE] providercache.fillMetaCache: using cached result from previous scan of .terraform/plugins
2020/07/15 11:25:35 [DEBUG] checking for provisioner in "."
2020/07/15 11:25:35 [DEBUG] checking for provisioner in "/usr/local/bin"
2020/07/15 11:25:35 [INFO] Failed to read plugin lock file .terraform/plugins/linux_amd64/lock.json: open .terraform/plugins/linux_amd64/lock.json: no such file or directory
2020/07/15 11:25:35 [TRACE] Meta.Backend: backend <nil> does not support operations, so wrapping it in a local backend
2020/07/15 11:25:35 [TRACE] backend/local: CLI option -state is overriding state path to terraform.tfstate
2020/07/15 11:25:35 [TRACE] backend/local: state manager for workspace "cm-aws-sandbox-us-west-2-dev-sandbox" will:
 - read initial snapshot from terraform.tfstate
 - write new snapshots to terraform.tfstate
 - create any backup at terraform.tfstate.backup
2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: preparing to manage state snapshots at terraform.tfstate
2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: no previously-stored snapshot exists
2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: locking terraform.tfstate using fcntl flock
2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: writing lock metadata to .terraform.tfstate.lock.info
2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: reading latest snapshot from terraform.tfstate
2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: snapshot file has nil snapshot, but that's okay
2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: read nil snapshot

2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
Error: No such resource instance

The state currently contains no resource instances whatsoever. This may occur
if the configuration has never been applied or if it has recently been
destroyed.

2020/07/15 11:25:35 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
```

**Comment by jbardin** [comment]
Thanks @issacg 

I'm not sure what's going on here, since the errors don't really seem related, so I'm going to reopen this for now pending more investigation.

**Comment by egonbraun** [comment]
Maybe a silly question, but wouldn't these log lines from the above state be saying that it is using a **new** state instead of the one you expect? Therefore being empty and containing no resources?

```
2020/07/15 11:25:34 [TRACE] Meta.Backend: no config given or present on disk, so returning nil config
2020/07/15 11:25:34 [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2020/07/15 11:25:34 [DEBUG] New state was assigned lineage "f3ff8551-67fc-696b-d570-b3e951738451"
```

If that's true then the question I think should be why we are not checking the current state and using a new one.

**Comment by danieldreier** [comment]
@issacg based on your and @jbardin's conversation I believe that this is real, but I'm trying to reproduce it and have not been able to come up with a clear working reproduction test case. I've put my best attempt at it in https://github.com/danieldreier/terraform-issue-reproductions/tree/master/25535 and it would be super helpful if you could make a PR against that to show the problem you're encountering. I am using null resources to simplify testing. Is this something you can help with?

**Comment by issacg** [comment]
Hey Daniel,

I did one better.  I tried to find a git snapshot from approximately when I originally encountered the bug, including the vscode devcontainer that pinned the versions, in order to share.

Unfortunately, (or perhaps fortunately, depending on one's perspective ;)) I can't reproduce the issue myself either with something that at least highly resembles the original terraform code 

I'm closing this one.  Thanks for the valiant effort! 👍
